price-feeder/oracle/provider/gate.go

```diff
@@ -0,0 +1,525 @@
+		reconnectTimer:  time.NewTicker(gatePingCheck),
+		tickers:         map[string]GateTicker{},
+		candles:         map[string][]GateCandle{},
+		subscribedPairs: map[string]types.CurrencyPair{},
+	}
+	provider.wsClient.SetPongHandler(provider.pongHandler)
+
+	if err := provider.SubscribeCurrencyPairs(pairs...); err != nil {
+		return nil, err
+	}
+
+	go provider.handleReceivedTickers(ctx)
+
+	return provider, nil
+}
+
+// GetTickerPrices returns the tickerPrices based on the saved map.
+func (p *GateProvider) GetTickerPrices(pairs ...types.CurrencyPair) (map[string]TickerPrice, error) {
+	tickerPrices := make(map[string]TickerPrice, len(pairs))
+
+	for _, currencyPair := range pairs {
+		price, err := p.getTickerPrice(currencyPair)
+		if err != nil {
+			return nil, err
+		}
+
+		tickerPrices[currencyPair.String()] = price
+	}
+
+	return tickerPrices, nil
+}
+
+// GetCandlePrices returns the candlePrices based on the saved map
+func (p *GateProvider) GetCandlePrices(pairs ...types.CurrencyPair) (map[string][]CandlePrice, error) {
+	candlePrices := make(map[string][]CandlePrice, len(pairs))
+
+	for _, currencyPair := range pairs {
+		gp := currencyPairToGatePair(currencyPair)
+		price, err := p.getCandlePrices(gp)
+		if err != nil {
+			return nil, err
+		}
+
+		candlePrices[currencyPair.String()] = price
+	}
+
+	return candlePrices, nil
+}
+
+func (p *GateProvider) getCandlePrices(key string) ([]CandlePrice, error) {
+	p.mtx.RLock()
+	defer p.mtx.RUnlock()
+
+	candles, ok := p.candles[key]
+	if !ok {
+		return []CandlePrice{}, fmt.Errorf("failed to get candle prices for %s", key)
+	}
+
+	candleList := []CandlePrice{}
+	for _, candle := range candles {
+		cp, err := candle.toCandlePrice()
+		if err != nil {
+			return []CandlePrice{}, err
+		}
+		candleList = append(candleList, cp)
+	}
+	return candleList, nil
+}
+
+// SubscribeCurrencyPairs subscribe all currency pairs into ticker and candle channels.
+func (p *GateProvider) SubscribeCurrencyPairs(cps ...types.CurrencyPair) error {
+	if err := p.subscribeTickers(cps...); err != nil {
+		return err
+	}
+	if err := p.subscribeCandles(cps...); err != nil {
+		return err
+	}
+	p.setSubscribedPairs(cps...)
+	return nil
+}
+
+// subscribeTickers subscribes to the ticker channels for all pairs at once.
+func (p *GateProvider) subscribeTickers(cps ...types.CurrencyPair) error {
+	topics := []string{}
+
+	for _, cp := range cps {
+		topics = append(topics, currencyPairToGatePair(cp))
+	}
+
+	tickerMsg := newGateTickerSubscription(topics...)
+	if err := p.subscribeTickerPairs(tickerMsg); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// subscribeCandles subscribes to the candle channels for all pairs one-by-one.
+// The gate API currently only supports subscribing to one kline market at a time.
+//
+// REF: https://www.gate.io/docs/websocket/#kline-subscription
+func (p *GateProvider) subscribeCandles(cps ...types.CurrencyPair) error {
```

Comment:
Just want to highlight this. Per gate's docs : 

```
Can only subscribe to one market at the same time, market list is not supported currently. 
For multiple subscriptions, only the last one takes effect.
```

tldr; we can only use gate for **one** currency pair (which will have to be umee) until they support multiple. 
We could do some strict config validation for this, but seems like we should just add a log warning. wdyt @alexanderbez ?

Reply:
lol what a shitty API. My proposal to actually add a (temporary) error to the config validation where an error is thrown if gate is used for more than one pair. WDYT?

Reply:
Yup! Agreed